scripts/py_matter_yamltests/matter_yamltests/constraints.py

```diff
@@ -27,14 +27,17 @@ def __init__(self, message):
 class BaseConstraint(ABC):
     '''Constraint Interface'''
 
-    def __init__(self, types: list, is_null_allowed: bool = False):
+    def __init__(self, types: list):
         '''An empty type list provided that indicates any type is accepted'''
         self._types = types
-        self._is_null_allowed = is_null_allowed
 
     def is_met(self, value, value_type_name):
         if value is None:
-            return self._is_null_allowed
+            # To be in alignment with codegen, whenever something that is optional
+            # is not provided we are supposed to not enforce the constraint. The
+            # only constraion that we are supposed to check is has value, which
+            # overrides is_met.
```

Comment:
If something is _optional_ and _nullable_, then the `notValue: null` constraint should in fact be enforced if the value is present (as in, it should fail if present and null).

But also, what happens with `notValue: null` with this PR for something that is _not_ optional?  It looks to me like it will just skip enforcement and pass it, which is not right....

Reply:
I wonder how python code represents null vs missing. Do we have a null placeholder or `is_present` equivalent or something?

I.e. in this context does `value is Null` mean the value was missing (because maybe optional) or does it mean the value was present with a null value in it.

Reply:
Looking at repl documentation:

https://deepnote.com/viewer/github/project-chip/connectedhomeip/blob/master/docs/guides/repl/Matter_Basic_Interactions.ipynb

it says `an optional field that is not present has a value of None` and it looks like Null is a separate value (that is not none)

Reply:
Sorry for the delayed reply here. In chip-repl an optional field that is not present has a value of None, a null value is `NullValue`. We don't have this same level of granularity for the common parser just yet. I will work with Vivien to see if the chip-tool based method of running can differentiate. If they can we can update the common parser to do so